### PR TITLE
[docker] feat: add Ascend GDN Dockerfiles

### PR DESCRIPTION
**Pull request by xssty123**
### What does this PR do?

> Add three new CANN 9.0.0 / torch-npu 2.10.0.post2 Dockerfiles for running
> Qwen3.5 GatedDeltaNet (GDN) on Ascend A2 / 910B amd64, A2 / 910B arm64,
> and A3 arm64.
>
> The GDN NPU backends were introduced in #879 and #924. These Dockerfiles
> provide the two additional libraries required by those backends:
>
> - `triton-ascend==3.2.1` for the Triton-based NPU backend from #879.
> - `fla_npu`, built from `flash-linear-attention-npu` commit `c2e3d83f`, for
>   the AscendC backend from #924.
>
> All three files are added alongside the existing CANN 9.0.0 Dockerfiles.
> No existing Dockerfile is replaced or modified.

### Checklist Before Starting

- Search for relative PRs/issues and link here: #879 and #924.
- PR title follows `[{modules}] {type}: {description}` format.

### Test

> GitHub Actions build validation:
>
> - A2 / 910B amd64 image built successfully.
> - A2 / 910B arm64 image built successfully.
> - A3 arm64 image built successfully.
>
> Physical NPU validation:
>
> - The A2 images and GDN runtime were validated successfully on physical A2
>   hardware.
> - The A3 image and GDN runtime were validated successfully on physical A3
>   hardware.

### API and Usage Example

> N/A. This PR adds Dockerfiles and does not change any Python API or
> configuration field.

### Design & Code Changes

> - Add separate GDN-ready Dockerfiles for A2 amd64, A2 arm64, and A3 arm64,
>   while keeping the existing Ascend Dockerfiles unchanged.
> - Install `torch-npu==2.10.0.post2`, `triton-ascend==3.2.1`.
> - Build and install the `fla_npu` AscendC operators and PyTorch bindings from
>   pinned commit `c2e3d83f`.
> - Verify the Triton Ascend backend import during the image build before
>   compiling and installing `fla_npu`.
> - Update the Ascend image overview, supported-tag links, and A2/A3 build
>   commands for the new GDN-ready Dockerfiles.

### Checklist Before Submitting

- Read the [Contribute Guide](https://github.com/ByteDance-Seed/VeOmni/blob/main/CONTRIBUTING.md).
- Applied pre-commit checks: `git diff --check` passed, and all Docker builds
  passed in GitHub Actions.
- Added/updated documentation: updated the Ascend image overview, supported
  tags, and A2/A3 build guides for the new Dockerfiles.
- If `tasks/` training scripts were moved or renamed: N/A. No training script
  is changed.
- Added tests to CI workflow (or explained why not feasible): no workflow is
  included in this Dockerfile-only PR; all three Dockerfiles were validated by
  build-only GitHub Actions before submission.

## Summary by CodeRabbit

* **New Features**
  * Added Ascend 910B and A3 Docker images for ARM64 and x86 platforms.
  * Added NPU acceleration support with `torch-npu`, `triton-ascend`, and `fla_npu`.
  * Included TorchCodec support in applicable images.
  * Added GDN-ready configurations with architecture-specific dependencies and build options.

* **Documentation**
  * Updated Ascend build guides, supported tags, image links, architectures, and included components.
  * Added Qwen3-VL smoke tests and Qwen3.5 GDN training examples.

### CODE REVIEW

File: docker/ascend/Dockerfile.ascend_9.0.0_torch_npu2.10.0.post2_910b.arm [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+
+COPY . .
+
+RUN pip install -e .[npu_aarch64]
+
+# Align the parquet/data stack with the uv lockfile. pip install -e (unlike x86's
+# uv sync --locked) ignores the lockfile and floats pyarrow/pandas to latest; pyarrow
+# 25 / pandas 3.x SEGFAULT reading parquet on aarch64 (datasets is pinned to 2.21, which
+# targets the older pyarrow ABI). Pin to the versions the x86 lockfile uses.
+RUN pip install "pyarrow==21.0.0" "pandas==2.3.2"
```
[the remark this answers is not in the document]
Comment (xssty123) [reply docker/ascend/Dockerfile.ascend_9.0.0_torch_npu2.10.0.post2_910b.arm]: Fixed in 9648fb33 by quoting the editable extras specifier. The same fix was applied to the A3 Dockerfile for consistency.

File: docker/ascend/Dockerfile.ascend_9.0.0_torch_npu2.10.0.post2_910b.arm [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+# dep, confirmed via `pip install triton-ascend==3.2.1 --dry-run` on an aarch64 host.
+# pip tolerates the osinfra index behavior here; x86 uses uv with the Huawei Cloud mirror
+# because uv rejects wrong-name metadata returned by osinfra for transitive lookups.
+RUN (pip uninstall -y triton triton-ascend || true) && \
+    pip install triton==3.5.0 && \
+    pip install triton-ascend==3.2.1 \
+        --extra-index-url=https://triton-ascend.osinfra.cn/pypi/simple \
+        --trusted-host triton-ascend.osinfra.cn && \
+    echo "Verifying triton-ascend patched libtriton.so ..." && \
+    python -c "from triton._C.libtriton import ascend"
```
[the remark this answers is not in the document]
Comment (xssty123) [reply docker/ascend/Dockerfile.ascend_9.0.0_torch_npu2.10.0.post2_910b.arm:77]: Fixed in d9735d1f. Removed --trusted-host and switched the ARM image to the HTTPS Huawei Cloud Ascend package index. The equivalent build passed the A2 arm64 build-only Action on the test branch.

File: docker/ascend/Dockerfile.ascend_9.0.0_torch_npu2.10.0.post2_910b.arm [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+# torch binding wheel is force-reinstalled with --no-deps (else pip pulls a CUDA torch).
+RUN git clone https://github.com/flashserve/flash-linear-attention-npu /tmp/fla-npu && \
+    cd /tmp/fla-npu && git checkout c2e3d83f && \
+    source /usr/local/Ascend/ascend-toolkit/set_env.sh && \
+    python scripts/check_npu_env.py --build-only && \
+    bash build.sh --soc=ascend910b --pkg --vendor_name=fla_npu && \
+    bash build_out/*.run --quiet && \
+    cd torch_custom/fla_npu && bash build.sh && \
+    pip install --no-deps --force-reinstall dist/*.whl && \
+    cd / && rm -rf /tmp/fla-npu
```
[the remark this answers is not in the document]
Comment (xssty123) [reply docker/ascend/Dockerfile.ascend_9.0.0_torch_npu2.10.0.post2_910b.arm:89]: Fixed in 9648fb33. Updated the A2 build guide, image overview, and supported-tag links to reference the new GDN-ready ARM Dockerfile.

File: docker/ascend/Dockerfile.ascend_9.0.0_torch_npu2.10.0.post2_910b.x86 [hunk 4]
```diff
@@ -0,0 +1,143 @@
+# docker hub: https://hub.docker.com/layers/ascendai/cann/9.0.0-910b-ubuntu22.04-py3.11/images/sha256-latest
+# cann web: https://www.hiascend.com/developer/ascendhub/detail/17da20d1c2b6493cb38765adeba85884
+# git repo: https://gitcode.com/Ascend/pytorch
+FROM swr.cn-south-1.myhuaweicloud.com/ascendhub/cann:9.0.0-910b-ubuntu22.04-py3.11
```
[the remark this answers is not in the document]
Comment (xssty123) [reply docker/ascend/Dockerfile.ascend_9.0.0_torch_npu2.10.0.post2_910b.x86:4]: Fixed in 9648fb33. Updated the image overview, supported tags, and A2 build commands to reference the new GDN-ready 910B x86 Dockerfile.

File: docker/ascend/Dockerfile.ascend_9.0.0_torch_npu2.10.0.post2_910b.x86 [hunk 5]
```diff
@@ -0,0 +1,143 @@
...
+
+# Install uv
+COPY --from=ghcr.io/astral-sh/uv:0.11.16 /uv /uvx /bin/
+ENV PATH="/bin/:$PATH"
+ENV UV_CACHE_DIR=/home/tiger/.cache/uv
+
+# uv sync
+WORKDIR /app
+COPY --chown=tiger:tiger . ./
+RUN uv sync --python 3.11 --locked --extra npu --allow-insecure-host github.com --allow-insecure-host objects.githubusercontent.com --allow-insecure-host download-r2.pytorch.org
```
[the remark this answers is not in the document]
Comment (xssty123) [reply docker/ascend/Dockerfile.ascend_9.0.0_torch_npu2.10.0.post2_910b.x86]: Fixed across d9735d1f and 9648fb33. uv sync now uses --system-certs with no insecure-host exceptions, and the Ascend documentation now references the new x86 Dockerfile. The equivalent amd64 build passed the build-only Action on the test branch.

File: docker/ascend/Dockerfile.ascend_9.0.0_torch_npu2.10.0.post2_a3 [hunk 6]
```diff
@@ -0,0 +1,119 @@
+FROM swr.cn-south-1.myhuaweicloud.com/ascendhub/cann:9.0.0-a3-ubuntu22.04-py3.11
```
[the remark this answers is not in the document]
Comment (xssty123) [reply docker/ascend/Dockerfile.ascend_9.0.0_torch_npu2.10.0.post2_a3:1]: Fixed in 9648fb33. Updated the A3 build guide with the new GDN-ready Dockerfile, build command, and PIP_INDEX/APT_SOURCE inputs; overview and supported-tag links were updated as well.

File: docker/ascend/Dockerfile.ascend_9.0.0_torch_npu2.10.0.post2_a3 [hunk 7]
```diff
@@ -0,0 +1,119 @@
...
+# dep, confirmed via `pip install triton-ascend==3.2.1 --dry-run` on an aarch64 host.
+# pip tolerates the osinfra index behavior here; x86 uses uv with the Huawei Cloud mirror
+# because uv rejects wrong-name metadata returned by osinfra for transitive lookups.
+RUN (pip uninstall -y triton triton-ascend || true) && \
+    pip install triton==3.5.0 && \
+    pip install triton-ascend==3.2.1 \
+        --extra-index-url=https://triton-ascend.osinfra.cn/pypi/simple \
+        --trusted-host triton-ascend.osinfra.cn && \
+    echo "Verifying triton-ascend patched libtriton.so ..." && \
+    python -c "from triton._C.libtriton import ascend"
```
[the remark this answers is not in the document]
Comment (xssty123) [reply docker/ascend/Dockerfile.ascend_9.0.0_torch_npu2.10.0.post2_a3:89]: Fixed in d9735d1f. Removed --trusted-host and switched to the HTTPS Huawei Cloud Ascend package index. The equivalent A3 arm64 build passed the build-only Action on the test branch.

**Comment by xssty123** [comment]
Review follow-up:

- Addressed all seven inline findings in `d9735d1f` and `9648fb33`: TLS verification, quoted extras, and A2/A3 documentation links/build commands.
- The repeated pip bootstrap sequence was intentionally left unchanged. It is inherited from the existing `Dockerfile.ascend_9.0.0_a2.arm` / `Dockerfile.ascend_9.0.0_a3` baseline and is unrelated to the GDN additions; consolidating those layers would broaden this PR and require separate image regression validation.
- The TLS-equivalent A2 amd64, A2 arm64, and A3 arm64 Dockerfiles all passed the private build-only Actions before these commits were applied to the PR branch.

File: docs/hardware_support/AscendDockerUsage/build_a2_docker.md [hunk 8]
```diff
@@ -189,6 +192,24 @@ bash train.sh tasks/train_vlm.py configs/multimodal/qwen3_vl/qwen3_vl_dense.yaml
...
+    --data.train_path ./configs/multimodal/data/tulu_sharegpt4v_llavavideo.yaml \
+    --model.ops_implementation.rms_norm_gated_implementation npu \
+    --model.ops_implementation.causal_conv1d_implementation npu \
+    --model.ops_implementation.chunk_gated_delta_rule_implementation npu_ascendc \
+    --train.max_steps 20
+```
+
+See the [Qwen3.5 training guide](../../examples/qwen3_5.md#start-training-on-npu)
+for backend details and the Qwen3.5 MoE example.
+
```
[the remark this answers is not in the document]
Comment (xssty123) [reply docs/hardware_support/AscendDockerUsage/build_a2_docker.md:219]: Keeping the dense Qwen3.5 configuration intentionally. Dense versus MoE describes the FFN architecture; the dense Qwen3.5 model still contains GatedDeltaNet linear-attention layers. This example explicitly selects the NPU and AscendC GDN backends, while the separate MoE example remains linked below. Switching this command to the MoE configuration would remove the intended dense GDN example.

File: docker/ascend/Dockerfile.ascend_9.0.0_torch_npu2.10.0.post2_910b.arm [hunk 9]
```diff
@@ -0,0 +1,104 @@
...
+    pip config --user set global.progress_bar off && \
+    python -m pip install --upgrade pip
+
+RUN pip3 install --no-cache-dir -U pip setuptools requests
+
+WORKDIR /app
+
+COPY . .
+
+RUN pip install -e ".[npu_aarch64]"
```
[the remark this answers is not in the document]
Comment (xssty123) [reply docker/ascend/Dockerfile.ascend_9.0.0_torch_npu2.10.0.post2_910b.arm]: Not changing the diffusers pin in this PR. diffusers==0.37.0 is inherited from main and is shared by the GPU, NPU, and NPU aarch64 extras; this PR does not change dependency metadata. We agree that the security advisory should be addressed, but that requires a coordinated pyproject.toml and uv.lock update plus cross-platform validation, so it should be handled in a dedicated dependency/security PR.

File: docs/hardware_support/AscendDockerUsage/build_a3_docker.md [hunk 10]
```diff
@@ -19,26 +19,28 @@ docker pull --platform=arm64 swr.cn-south-1.myhuaweicloud.com/ascendhub/cann:9.0
 ```
 
 ## Step 2: Build the Custom Image
-Build the VeOmni Ascend A3 image using the provided Dockerfile.
+Build the GDN-ready VeOmni Ascend A3 image using the provided Dockerfile. It includes `torch-npu==2.10.0.post2`, `triton-ascend`, and `fla_npu`. The existing `Dockerfile.ascend_9.0.0_a3` remains available for the general-purpose image.
 
 **Note:** Proxy settings are optional and only needed if your server requires proxy access to the internet. Remove the proxy arguments if not needed.
 
+The GDN-ready Dockerfile also accepts `PIP_INDEX` and `APT_SOURCE` build arguments. `APT_SOURCE` must point to an Ubuntu Ports mirror.
```
[the remark this answers is not in the document]
Comment (xssty123) [reply docs/hardware_support/AscendDockerUsage/build_a3_docker.md:26]: This is already addressed in the current documentation. The generic command is explicitly labeled General Qwen3-VL Smoke Test, and a separate Qwen3.5 GDN Training Example selects the NPU and AscendC GDN backends. No additional command replacement is needed.

**Comment by xssty123** [comment]
Regarding the outside-diff TorchCodec pinning suggestion: this PR intentionally keeps `git checkout v0.5.0`. The checkout is inherited from the existing Ascend Dockerfiles on `main`, and `v0.5.0` currently resolves to the suggested commit (`93cc0ade58be5e424fc17489f1ba50b7b530d00b`), matching the source used by the validated builds. Converting all Ascend Dockerfiles to immutable commit pins is reasonable hardening, but should be a separate consistency change with its own rebuild validation.

**Review by Crystal-jiang — APPROVED** [review]
LGTM, ready to merge.
